File: csharp/Minesweeper.Engine/Point.cs
namespace MineSweeper.Engine
{
    // Represents a point or a position on the game grid
    public class Point(int row, int col)
    {
        public int Row { get; private set; } = row;
        public int Col { get; private set; } = col;

        public override string ToString()
        {
            return $"({Row}, {Col})";
        }

        public static Point? TryParse(string? input, int gridSize)
        {
            if (string.IsNullOrWhiteSpace(input))
                return null; // Input cannot be null or empty.

            input = input.ToUpper().Trim();
            if (input.Length < 2 || input.Length > 3)
                return null; // Invalid input format.

            int row = input[0] - 'A';
            int col = int.Parse(input.Substring(1)) - 1;

            if (row < 0 || row >= gridSize || col < 0 || col >= gridSize)
                return null; // Input out of grid bounds.

            return new Point(row, col);
        }
    }

}

File: kotlin/src/main/kotlin/engine/Point.kt
package org.minesweeper.engine

/**
 * Represents a point or a position on the game grid
 */
data class Point(val row: Int, val col: Int) {
    override fun toString(): String = "($row, $col)"

    companion object {
        fun tryParse(input: String?, gridSize: Int): Point? {
            input ?: return null // Input cannot be null or empty.
            val trimmed = input.uppercase().trim()
            if (trimmed.length !in 2..3) return null // Invalid input format.

            val row = trimmed[0] - 'A'
            val col = trimmed.substring(1).toInt() - 1

            if (row !in 0 until gridSize || col !in 0 until gridSize) return null // Input out of grid bounds.

            return Point(row, col)
        }
    }
}